### DISCUSSION TITLE
Pytest fixture hooks for dialect developers

### DISCUSSION THREAD

**Discussion by andersbogsnes** [sqlalchemy/sqlalchemy#11718] [Ideas]
When writing a SQLAlchemy dialect, it would be useful to have the ability to use pytest fixtures to do configuration of the SQLAlchemy testsuite such as setting the engine or the db_url. That way, I can create a Docker-based database through something like Testcontainers and pass the dynamically generated endpoint to the SQLAlchemy pytest plugin. 

In terms of prior art, see https://anyio.readthedocs.io/en/stable/testing.html for how we can configure which asyncio backend to use, or https://pytest-alembic.readthedocs.io/en/latest/setup.html#setting-up-fixtures which defines the alembic configuration or the engine to use. 

For dialect development purpose, I'd like to have an `sqlalchemy_engine` fixture, similar to how `pytest-alembic` does today, to be exposed which I could override to set the engine from within my testsuite.

Other ideas for overridable fixtures could be MetaData and the various test tables definitions

**Comment by CaselIT** [comment]
Hi

It is currently not possible to support a fixture like that, since sqlalchemy and alembic test suites need to know what url are available while setting up the test suite, for example at pytest_modifyitems time. This is because the test suites can use multiple urls at the same time by dynamically creating test classes for each url.

While it may be possible to change the current logic to support this use case it would require a very large refactoring of the test suite, that may also require changes to 3rd party dialects suites.

Note the linked package is used to test user code that uses alembic, not the library itself.

**Comment by andersbogsnes** [comment]
Ok, thanks for the feedback - is there any mechanism within the current test suite to provide such hooks? I see PyAthena is using some kind of registry from `sqlalchemy.testing.provision`, but I can't find documentation for how it's suppposed to work. https://github.com/laughingman7743/PyAthena/blob/master/tests/sqlalchemy/conftest.py

When i try to use those decorators, they don't seem to get executed before the creation of the test tables

**Reply by CaselIT** [reply]
I don't think it's documented when each function is called in a clear way.

Some of the provision functions are called before running the tests, like generate_driver_url butthe default implementation is usually fine for it.
